### PR TITLE
Introduce hp::Collection

### PR DESCRIPTION
**Pull request by peterrum**
Preparation for #11564 so that we only need to implement the internal logic once.

### CODE REVIEW

File: include/deal.II/hp/q_collection.h [hunk 1]
```diff
@@ -235,12 +196,12 @@ namespace hp
...
   QCollection<dim>::operator==(const QCollection<dim> &q_collection) const
   {
-    const unsigned int n_quadratures = size();
+    const unsigned int n_quadratures = this->size();
     if (n_quadratures != q_collection.size())
       return false;
 
     for (unsigned int i = 0; i < n_quadratures; ++i)
-      if (!(*quadratures[i] == q_collection[i]))
+      if ((this->operator[](i) == q_collection[i]) == false)
```
Comment (bangerth) [inline include/deal.II/hp/q_collection.h:204]:
```suggestion
      if (this->operator[](i) != q_collection[i])
```
:-)

**Review by bangerth — APPROVED** [review]
I'll merge to allow you to make progress, but please address the one issue in a follow-up patch.

File: include/deal.II/hp/q_collection.h (hunk 1 above)
Comment (peterrum) [reply include/deal.II/hp/q_collection.h:204]: I'll do (tomorrow).

**Pull request merged**